### PR TITLE
[SPARK-28228][SQL] Change the default behavior for name conflict in nested WITH clause

### PR DESCRIPTION
**Pull request by xuanyuanking**
### What changes were proposed in this pull request?
This is a follow-up for #25029, in this PR we throw an AnalysisException when name conflict is detected in nested WITH clause. In this way, the config `spark.sql.legacy.ctePrecedence.enabled` should be set explicitly for the expected behavior.

### Why are the changes needed?
The original change might risky to end-users, it changes behavior silently.

### Does this PR introduce any user-facing change?
Yes, change the config `spark.sql.legacy.ctePrecedence.enabled` as optional.

### How was this patch tested?
New UT.

### CODE REVIEW

**Comment by xuanyuanking** [comment]
cc @cloud-fan

**Comment by xuanyuanking** [comment]
retest this please.

**Comment by HyukjinKwon** [comment]
retest this please

File: docs/sql-migration-guide.md [hunk 1]
```diff
@@ -103,7 +103,7 @@ license: |
...
 
-  - Since Spark 3.0, substitution order of nested WITH clauses is changed and an inner CTE definition takes precedence over an outer. In version 2.4 and earlier, `WITH t AS (SELECT 1), t2 AS (WITH t AS (SELECT 2) SELECT * FROM t) SELECT * FROM t2` returns `1` while in version 3.0 it returns `2`. The previous behaviour can be restored by setting `spark.sql.legacy.ctePrecedence.enabled` to `true`.
+  - Since Spark 3.0, an AnalysisException will happen by default in case of nested WITH clauses, it forces the users to choose the specific substitution order they wanted, which is controlled by `spark.sql.legacy.ctePrecedence.enabled`. If set to false, inner CTE definitions take precedence over outer definitions. For example, set the config to `false`, `WITH t AS (SELECT 1), t2 AS (WITH t AS (SELECT 2) SELECT * FROM t) SELECT * FROM t2` returns `1` returns `2`, while setting it to `true`, the result is `1` which is the behavior in version 2.4 and earlier.
```
Comment (cloud-fan) [inline docs/sql-migration-guide.md]: Since Spark 3.0, Spark throws an AnalysisException if name conflict is detected in the nested WITH clause.

Comment (cloud-fan) [reply docs/sql-migration-guide.md]: returns `1` returns `2` ?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 2]
```diff
@@ -2714,6 +2714,8 @@ class SQLConf extends Serializable with Logging {
 
   def csvFilterPushDown: Boolean = getConf(CSV_FILTER_PUSHDOWN_ENABLED)
 
+  def legacyCTEPrecedenceEnabled: Option[Boolean] = getConf(LEGACY_CTE_PRECEDENCE_ENABLED)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: I think it's not encouraged to create a method now. We should call `getConf` directly.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala [hunk 3]
```diff
@@ -17,24 +17,57 @@
...
 import org.apache.spark.sql.internal.SQLConf
-import org.apache.spark.sql.internal.SQLConf.LEGACY_CTE_PRECEDENCE_ENABLED
 
 /**
  * Analyze WITH nodes and substitute child plan with CTE definitions.
  */
 object CTESubstitution extends Rule[LogicalPlan] {
   def apply(plan: LogicalPlan): LogicalPlan = {
-    if (SQLConf.get.getConf(LEGACY_CTE_PRECEDENCE_ENABLED)) {
+    if (SQLConf.get.legacyCTEPrecedenceEnabled.isEmpty) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala]: nit:
```
val isLegacy = SQLConf.get.getConf(LEGACY_CTE_PRECEDENCE_ENABLED)
if (isLegacy.isEmpty) ...
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala [hunk 4]
```diff
@@ -17,24 +17,57 @@
...
+      if (hasNestedCTE(plan, inTraverse = false)) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala]: we should only fail if there are name conflicts. It's a bad UX if we blindly forbid nested CTE by default.

Comment (xuanyuanking) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala]: Thanks for the advice, done in a55e82d and change tests correspondingly.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala (hunk 3 above)
Comment (xuanyuanking) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala]: Thanks, done in a55e82d.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 2 above)
Comment (xuanyuanking) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Thanks, deleted in a55e82d.

File: docs/sql-migration-guide.md (hunk 1 above)
Comment (xuanyuanking) [reply docs/sql-migration-guide.md]: Thanks for the advice, done in a55e82d.

File: docs/sql-migration-guide.md [hunk 5]
```diff
@@ -103,7 +103,7 @@ license: |
...
 
-  - Since Spark 3.0, substitution order of nested WITH clauses is changed and an inner CTE definition takes precedence over an outer. In version 2.4 and earlier, `WITH t AS (SELECT 1), t2 AS (WITH t AS (SELECT 2) SELECT * FROM t) SELECT * FROM t2` returns `1` while in version 3.0 it returns `2`. The previous behaviour can be restored by setting `spark.sql.legacy.ctePrecedence.enabled` to `true`.
+  - Since Spark 3.0, Spark throws an AnalysisException if name conflict is detected in the nested WITH clause. It forces the users to choose the specific substitution order they wanted, which is controlled by `spark.sql.legacy.ctePrecedence.enabled`. If set to false, inner CTE definitions take precedence over outer definitions. For example, set the config to `false`, `WITH t AS (SELECT 1), t2 AS (WITH t AS (SELECT 2) SELECT * FROM t) SELECT * FROM t2` returns `2`, while setting it to `true`, the result is `1` which is the behavior in version 2.4 and earlier.
```
Comment (cloud-fan) [inline docs/sql-migration-guide.md]: `If set to false (which is recommended)`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala [hunk 6]
```diff
@@ -28,13 +29,55 @@ import org.apache.spark.sql.internal.SQLConf.LEGACY_CTE_PRECEDENCE_ENABLED
  */
 object CTESubstitution extends Rule[LogicalPlan] {
   def apply(plan: LogicalPlan): LogicalPlan = {
-    if (SQLConf.get.getConf(LEGACY_CTE_PRECEDENCE_ENABLED)) {
+    val isLegacy = SQLConf.get.getConf(LEGACY_CTE_PRECEDENCE_ENABLED)
+    if (isLegacy.isEmpty) {
+      if (detectConflictInNestedCTE(plan, inTraverse = false)) {
+        throw new AnalysisException("Name collision in nested CTE was founded in current plan, " +
+          s"please select the CTE precedence via ${LEGACY_CTE_PRECEDENCE_ENABLED.key}, " +
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala]: `Please set ${LEGACY_CTE_PRECEDENCE_ENABLED.key} to false so that name defined in inner CTE takes precedence. See more details in SPARK-28228.`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala [hunk 7]
```diff
@@ -28,13 +29,55 @@ import org.apache.spark.sql.internal.SQLConf.LEGACY_CTE_PRECEDENCE_ENABLED
  */
 object CTESubstitution extends Rule[LogicalPlan] {
   def apply(plan: LogicalPlan): LogicalPlan = {
-    if (SQLConf.get.getConf(LEGACY_CTE_PRECEDENCE_ENABLED)) {
+    val isLegacy = SQLConf.get.getConf(LEGACY_CTE_PRECEDENCE_ENABLED)
+    if (isLegacy.isEmpty) {
+      if (detectConflictInNestedCTE(plan, inTraverse = false)) {
+        throw new AnalysisException("Name collision in nested CTE was founded in current plan, " +
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala]: it's good if we can show the conflicting names.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala [hunk 8]
```diff
@@ -28,13 +29,55 @@ import org.apache.spark.sql.internal.SQLConf.LEGACY_CTE_PRECEDENCE_ENABLED
...
+  private def detectConflictInNestedCTE(
+      plan: LogicalPlan,
+      inTraverse: Boolean,
+      cteNames: Set[String] = Set.empty): Boolean = {
+    plan.foreach {
+      case w @ With(child, relations) =>
+        val newNames = relations.map {
+          case (cteName, _) =>
+            if (cteNames.contains(cteName)) {
+              return true
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala]: we can throw exception here so that we know which name is conflicting.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala [hunk 9]
```diff
@@ -28,13 +29,55 @@ import org.apache.spark.sql.internal.SQLConf.LEGACY_CTE_PRECEDENCE_ENABLED
...
-      traverseAndSubstituteCTE(plan, false)
+      traverseAndSubstituteCTE(plan, inTraverse = false)
     }
   }
 
+  /**
+   * Check the plan to be traversed has naming conflicts in nested CTE or not, traverse through
+   * child, innerChildren and subquery for the current plan.
+   */
+  private def detectConflictInNestedCTE(
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala]: nit: `assertNoNameConflictsInCTE`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala [hunk 10]
```diff
@@ -28,13 +29,55 @@ import org.apache.spark.sql.internal.SQLConf.LEGACY_CTE_PRECEDENCE_ENABLED
  */
 object CTESubstitution extends Rule[LogicalPlan] {
   def apply(plan: LogicalPlan): LogicalPlan = {
-    if (SQLConf.get.getConf(LEGACY_CTE_PRECEDENCE_ENABLED)) {
+    val isLegacy = SQLConf.get.getConf(LEGACY_CTE_PRECEDENCE_ENABLED)
+    if (isLegacy.isEmpty) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala]: nit:
```
if (isLegacy.isEmpty) {
  assertNoNameConflictsInCTE(...)
  traverseAndSubstituteCTE
} else if ...

```

File: sql/core/src/test/resources/sql-tests/inputs/cte-nonlegacy.sql [hunk 11]
```diff
@@ -0,0 +1,115 @@
+create temporary view t as select * from values 0, 1, 2 as t(id);
+create temporary view t2 as select * from values 0, 1 as t(id);
+
+-- CTE non-legacy substitution
+SET spark.sql.legacy.ctePrecedence.enabled=false;
+
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/cte-nonlegacy.sql]: we can use `--IMPORT cte.sql` to include the existing test

File: sql/core/src/test/resources/sql-tests/inputs/cte-nonlegacy.sql [hunk 12]
```diff
@@ -0,0 +1,115 @@
+create temporary view t as select * from values 0, 1, 2 as t(id);
+create temporary view t2 as select * from values 0, 1 as t(id);
+
+-- CTE non-legacy substitution
+SET spark.sql.legacy.ctePrecedence.enabled=false;
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/cte-nonlegacy.sql]: use `--SET spark.sql.legacy.ctePrecedence.enabled=false`, so that this won't be treated as a query and appear in the result query.

File: docs/sql-migration-guide.md (hunk 5 above)
Comment (xuanyuanking) [reply docs/sql-migration-guide.md]: Thanks, done in c03920a.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala (hunk 6 above)
Comment (xuanyuanking) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala]: Thanks, done in c03920a.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala (hunk 7 above)
Comment (xuanyuanking) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala]: Copy that, done in c03920a.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala (hunk 8 above)
Comment (xuanyuanking) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala]: Thanks, done in c03920a.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala (hunk 9 above)
Comment (xuanyuanking) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala]: Thanks, done in c03920a.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala (hunk 10 above)
Comment (xuanyuanking) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala]: Thanks, done in c03920a.

File: sql/core/src/test/resources/sql-tests/inputs/cte-nonlegacy.sql [hunk 13]
```diff
@@ -0,0 +1,2 @@
+--SET spark.sql.legacy.ctePrecedence.enabled = false
+--IMPORT cte.sql
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/cte-nonlegacy.sql]: we should do the same thing for `cte-legacy.sql`. This can be done in a followup.

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/test/resources/sql-tests/inputs/cte-nonlegacy.sql (hunk 13 above)
Comment (xuanyuanking) [reply sql/core/src/test/resources/sql-tests/inputs/cte-nonlegacy.sql]: Copy that, will do it later.

**Comment by dongjoon-hyun** [comment]
cc @peter-toth

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala [hunk 14]
```diff
@@ -28,10 +29,49 @@ import org.apache.spark.sql.internal.SQLConf.LEGACY_CTE_PRECEDENCE_ENABLED
...
+  private def assertNoNameConflictsInCTE(
+      plan: LogicalPlan,
+      inTraverse: Boolean,
+      cteNames: Set[String] = Set.empty): Unit = {
+    plan.foreach {
+      case w @ With(child, relations) =>
+        val newNames = relations.map {
+          case (cteName, _) =>
+            if (cteNames.contains(cteName)) {
+              throw new AnalysisException(s"Name $cteName is conflict in nested CTE. " +
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala]: Ur, `is conflict`? `conflict` is a verb or a noun, isn't it? Shall we revise a little?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 15]
```diff
@@ -2079,7 +2079,7 @@ object SQLConf {
     .internal()
     .doc("When true, outer CTE definitions takes precedence over inner definitions.")
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Now, it became `three-state conf`. Shall we mentioned more about `false` and `empty`?

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Sounds we should.

File: docs/sql-migration-guide.md [hunk 16]
```diff
@@ -103,7 +103,7 @@ license: |
...
 
-  - Since Spark 3.0, substitution order of nested WITH clauses is changed and an inner CTE definition takes precedence over an outer. In version 2.4 and earlier, `WITH t AS (SELECT 1), t2 AS (WITH t AS (SELECT 2) SELECT * FROM t) SELECT * FROM t2` returns `1` while in version 3.0 it returns `2`. The previous behaviour can be restored by setting `spark.sql.legacy.ctePrecedence.enabled` to `true`.
+  - Since Spark 3.0, Spark throws an AnalysisException if name conflict is detected in the nested WITH clause. It forces the users to choose the specific substitution order they wanted, which is controlled by `spark.sql.legacy.ctePrecedence.enabled`. If set to false (which is recommended), inner CTE definitions take precedence over outer definitions. For example, set the config to `false`, `WITH t AS (SELECT 1), t2 AS (WITH t AS (SELECT 2) SELECT * FROM t) SELECT * FROM t2` returns `2`, while setting it to `true`, the result is `1` which is the behavior in version 2.4 and earlier.
```
Comment (viirya) [inline docs/sql-migration-guide.md]: `Spark throws an AnalysisException ... by default.`?

**Review by viirya — APPROVED** [review]
Overall looks good. Only minor comment.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala [hunk 17]
```diff
@@ -28,10 +29,49 @@ import org.apache.spark.sql.internal.SQLConf.LEGACY_CTE_PRECEDENCE_ENABLED
...
+                s"Please set ${LEGACY_CTE_PRECEDENCE_ENABLED.key} to false so that name defined " +
+                "in inner CTE takes precedence. See more details in SPARK-28228.")
+            } else {
+              cteName
+            }
+        }.toSet
+        (w.innerChildren :+ child).foreach { p =>
```
Comment (peter-toth) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala]: This could be:
```
        child.transformExpressions {
          case e: SubqueryExpression =>
            assertNoNameConflictsInCTE(e.plan, inTraverse = true, cteNames ++ newNames)
            e
        }
        w.innerChildren.foreach { p =>
          assertNoNameConflictsInCTE(p, inTraverse = true, cteNames ++ newNames)
        }
```
If you check `CTE in subquery shadows outer` test cases you will see that legacy and new results are the same. We shouldn't give `AnalysisException` in those cases.

Comment (xuanyuanking) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala]: Thanks for the detailed checking! Yes, that makes sense, we should only give exception while legacy and non-legacy give different results. Done in 7249404

File: docs/sql-migration-guide.md (hunk 16 above)
Comment (xuanyuanking) [reply docs/sql-migration-guide.md]: Thanks, done in 7249404.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 15 above)
Comment (xuanyuanking) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Sure, done in 7249404.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala (hunk 14 above)
Comment (xuanyuanking) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala]: Thanks, rephrase it to `Name $name is ambiguous in nested CTE`. Done in 7249404.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 18]
```diff
@@ -2077,9 +2077,11 @@ object SQLConf {
 
   val LEGACY_CTE_PRECEDENCE_ENABLED = buildConf("spark.sql.legacy.ctePrecedence.enabled")
     .internal()
-    .doc("When true, outer CTE definitions takes precedence over inner definitions.")
+    .doc("When true, outer CTE definitions takes precedence over inner definitions. If set to " +
+      "false, inner CTE definitions take precedence. The default value is empty, which keep the " +
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Actually, `keep the behavior` is not correct because the 2.4.x behavior is `true`. The default value is new behavior in 3.0, isn't it?

Comment (xuanyuanking) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Yes, use `keep` here is ambiguous, thanks for pointing out! I want to express `always`.

Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: I'm not sure how exact and detailed this description needs to be, but
- if it is set to `false` then an inner CTE always takes precedence,
- if it is set to `true` then it complies with the (bit weird) 2.4 behaviour, where an outer CTE takes precedence in almost all of the cases (ie. inner CTE is in another CTE definition or in a subquery expression), except for the case where inner CTE is in a subquery.

**Review by peter-toth — APPROVED** [review]
LGTM, minor comment on legacy flag docs.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, all!
Merged to master/3.0.

**Comment by xuanyuanking** [comment]
Thanks all for the review.
